### PR TITLE
[Accelerate] [vDSP] Fill, Clear, and Generation Functions

### PR DESCRIPTION
**Pull request by FlexMonkey**
This PR contains a suite of Swift-friendly overlays to the vDSP vector fill, clear, and generation (including window and ramp) functions. The overlays simplify the API to the existing functions and accept collections that implement `_ContiguousCollection`.

cc: @moiseev @airspeedswift @stephentyrone

### CODE REVIEW

File: stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift
Comment (moiseev) [inline stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift]: It occurred to me, that maybe it is possible to make `CollectionOfOne` conform to the `_ContiguouslyStored` and avoid a singleton array allocation here.

Comment (moiseev) [reply stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift]: And... we can't. At least not that easily. `CollectionOfOne._element` is `internal`.

Comment (stephentyrone) [reply stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift]: We don't actually need to construct a singleton array here, right? Just a pointer to initialValue.

File: stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift
Comment (stephentyrone) [inline stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift]: @airspeedswift @moiseev `fill` feels like an API that we should be providing on MutableCollection?

**Review by moiseev — APPROVED** [review]

File: stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift
Comment (moiseev) [reply stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift]: @stephentyrone that would work too.

**Comment by moiseev** [comment]
@swift-ci Please test macOS platformn

**Comment by moiseev** [comment]
@swift-ci Please smoke test Linux platform

File: stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift
Comment (FlexMonkey) [reply stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift]: I've used a single element array as a shortcut, let me know if you think a pointer would be better form.

Comment (stephentyrone) [reply stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift]: A pointer would be a minor efficiency win. It's also just a bit cleaner.

Comment (airspeedswift) [reply stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift]: Should just be able to write:

```swift
var mutableTmp = initialValue
...
vDSP_vramp(&mutableTmp,
```

Comment (FlexMonkey) [reply stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift]: That does work, but it feels (at least to me) sort of clunky - we're not really mutating anything, just working round the vDSP API. Over in my [clipping PR](https://github.com/apple/swift/pull/23231/files), I use `withUnsafePointer`, I'll take your lead on a standardized approach.

File: stdlib/public/Darwin/Accelerate/ContiguousCollection.swift
Comment (jrose-apple) [inline stdlib/public/Darwin/Accelerate/ContiguousCollection.swift]: I'm very concerned about the proliferation of these protocols across multiple libraries.

File: stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift
Comment (airspeedswift) [reply stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift]: The `var` here means less mutation and more guaranteed materialization. Or that's how I rationalize it.

File: stdlib/public/Darwin/Accelerate/ContiguousCollection.swift
Comment (airspeedswift) [reply stdlib/public/Darwin/Accelerate/ContiguousCollection.swift]: Then you need to convince @DougGregor to like my pitch to add `ContiguousCollection` to the standard library :)

Comment (jrose-apple) [reply stdlib/public/Darwin/Accelerate/ContiguousCollection.swift]: That's…not a great answer. We (Apple SWE) try not to have copies of the same data structures in the actual Accelerate, Dispatch, CF, Foundation, etc; the overlays should follow the same rules.

Comment (stephentyrone) [reply stdlib/public/Darwin/Accelerate/ContiguousCollection.swift]: That's precisely why we want to have the protocol in the standard library?

In particular, Accelerate should not have an upwards dependency on Foundation.

Comment (jrose-apple) [reply stdlib/public/Darwin/Accelerate/ContiguousCollection.swift]: I happen to agree that we should have this in the stdlib, but that doesn't mean Accelerate should add this now. *If* Accelerate turns out to be the best place for it, Foundation should adopt that. Meanwhile, I don't want us to have several of these protocols *in-tree* and thus *serving as proposed ABI.*

Comment (stephentyrone) [reply stdlib/public/Darwin/Accelerate/ContiguousCollection.swift]: Ok, but what's @FlexMonkey or another API author supposed to do here? Do nothing and wait for the core team to get to considering the issue? Or put up draft API for review that can be updated once a decision is made? There's a substantial number of other issues to hammer out in review while we wait to see what the disposition of these protocols is going to be, and I don't think we should wait for that decision before we get started iterating on other API.

Comment (airspeedswift) [reply stdlib/public/Darwin/Accelerate/ContiguousCollection.swift]: The other consideration is that within Accelerate, the usage of this protocol is completely understood to mean a contiguous array of non-bridgeable types, bypassing concerns about lazily bridged arrays. This wouldn't apply if it were in Foundation or the standard library.

File: stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift
Comment (stephentyrone) [inline stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift]: Comment references a parameter that does not exist.

Comment (FlexMonkey) [reply stdlib/public/Darwin/Accelerate/vDSP_FillClearGenerate.swift]: Fixed for both single- and double-precision clear functions.

**Pull request closed without merging**
